### ISSUE TITLE
No services have been found

### ISSUE THREAD

**Issue by carher** [difi/Oxalis-AS4#65]
Initially it seemed the AS4 plugin worked fine with Oxalis-4.0.4, but we have since discovered that from time to time, something stops functioning. While AS2 keeps working fine, the AS4 endpoint may suddenly start responding almost immediately (much shorter response time than usual), and with response type text/html. From the HTML we see the following:

<title>CXF - Service list</title></head><body><span class=\"heading\">No services have been found.</span></body></html>

When we check the server log, we find the following information:
RESP_IN
Content-Type: text/html;charset=utf-8
ResponseCode: 200
ExchangeId: b2160f69-36bd-4026-a86b-f4cbf3f65b96
ServiceName: outbound-service
PortName: port
PortTypeName: DummyImpl
Headers: {content-type=text/html;charset=utf-8, Content-Length=363}
Payload: <!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"><HTML><HEAD><LINK type=\"text/css\" rel=\"stylesheet\" href=\"/as4/?stylesheet=1\"><meta http-equiv=\"content-type\" content=\"text/html; charset=UTF-8\"><title>CXF - Service list</title></head><body><span class=\"heading\">No services have been found.</span></body></html>

We also see org.apache.cxf.ws.policy.AssertionBuilderRegistryImpl log "No assertion builder for type {http://oxalis.difi.no/custom/security-policy}Basic128GCMSha256MgfSha256 registered."

While this happens, we still get the usual response from localhost:8080/as4. And, as mentioned, /as2 keeps functioning. If we restart the application, everything works fine for a while again.

Any ideas what may cause this problem?

**Comment by FrodeBjerkholt** [comment]
I haven't seen this before, but I will try to reproduce the problem tomorrow. 

Which version of Oxalis-AS4 did you use? Have you tried the latest release? Did you see a stacktrace or some other problem in the logs right before the endpoint stopped working?

**Comment by carher** [comment]
We are using 4.1.0-RC10, together with Oxalis 4.0.4.

We may try the released 4.1.0 of the plugin if you think that will make a difference.

We have so far not noticed anything in the log, apart from the aforementioned log statements from DummyImpl and AssertionBuilderRegistryImpl. What we suspect, though, is that it might be necessary to have had traffic on both AS2 and AS4 for the problem to appear.

**Comment by FrodeBjerkholt** [comment]
You can try the 4.1.1 (just released). 
I will try to reproduce the problem by sending a large number of AS2 and AS4 messages. Maybe there are some kind of memory leak somewhere? Do you have an approximate number of messages sent/received before the problem occur?

**Comment by FrodeBjerkholt** [comment]
I have now tested the 4.1.1 release on Tomcat 9.0.30 - sending 20.000 AS4 and 10.000 AS2 messages without reproducing the problem. Initially it used a lot of memory. Went up to 1.7 Gb, but soon stabilized around 0.3 Gb and then never exceeded 0.4 Gb. The Max heap setting was a bit large, so I will retry with a lower value.

**Comment by carher** [comment]
Ok.

I should add that we run on jetty-9.4.6.v20170531, using openjdk version "1.8.0_212".

**Comment by FrodeBjerkholt** [comment]
I did a new run with -Xms512M -Xmx1024M and then the memory consumption peaked at < 0.4 Gb.

**Comment by FrodeBjerkholt** [comment]
I have also tested the oxalis-server variant with the same result.

**Comment by gozus19p** [comment]
Hi all, I am experiencing the same issue.

I am currently using Oxalis-AS4 version "4.1.0-SNAPSHOT".
We configured our systems in order to support both AS2 and AS4.

You can find further details on our configuration below:
- Tomcat version 9.0.12
- JDK "OpenJDK Runtime Environment (build 1.8.0_222-b10)"
- Oxalis version 4.1.1-SNAPSHOT
- Oxalis-AS4 version "4.1.0-SNAPSHOT"

After restarting Tomcat server, the issue seems to disappear.
I didn't find the cause on this. Maybe some "cache logic"?

Let me know if I can help you further more. Thanks in advance.

**Comment by FrodeBjerkholt** [comment]
Please use the latest release. The RC versions are only release candidates. If you still experience the same issue, then I will have a new look at it.

**Comment by Intercent-ER** [comment]
The issue seems to actually be on the receiving (inbound) process, not on the outbound, so I suggest to test receiving a large number of files, instead of just sending them.

We are currently using Oxalis-AS4 4.1.1 (not the release candidate) and we're experiencing this issue in production environment

**Comment by FrodeBjerkholt** [comment]
Strange - I have used a single AP for receiving the large number of messages, without any problems.

**Comment by FrodeBjerkholt** [comment]
Can you check the logs and see if there is an interesting stacktrace or error right before the problem starts occuring?

**Comment by FrodeBjerkholt** [comment]
I just tested with tomcat-9.0.12 and it works as expected. To get anywhere with this issue I need logs showing what happened when the AP stopped working. It would be nice if you can email me a complete log file and a timestamp telling when the problem occur.

**Comment by FrodeBjerkholt** [comment]
Do you have an outbound servlet or similar on the same http server instance as the oxalis-inbound?
If so - can you try to run them on two separate instances? Like two tomcat servers on different ports?

When conformance testing we had problems running two instances of the same war on one tomcat server. The problems disappeared when we ran two tomcat servers instead.

**Comment by carher** [comment]
Yes, we have inbound and outbound on the same http server instance!

We are currently trying to upgrade the plugin version, but if that is no success, I will make sure to try splitting outbound and inbound. Sounds like a good idea!

**Comment by gozus19p** [comment]
Hi Frode, I am working with @Intercent-ER 

At the moment we have configured only one .war that includes AS2 and AS4 inbound servlets as well as outbound functionalities.
We run it using only one single Tomcat instance (9.0.12).

I could try to split .war into two different versions. The first version should include only outbound functionalities, the other one should include only inbound.

We just checked our logs, unfortunately all seems to go well until we reiceve an AS4 message.
We found no error log related to "No services found" issue.

**Comment by FrodeBjerkholt** [comment]
Let me know if splitting inbound and outbound works.

**Comment by zeko78** [comment]
Any updates on this? 

We're experiencing the same in production when running as2/as4 and outbound on the same tomcat instance (running as an Azure App Service).

We've moved AS4 to a separate instance a few hours ago, but got a few (unconfirmed) reports about the issue remaining. Self testing has not confirmed the issue after the split.

I was just interested in hearing about any development in this issue and if maybe @carher have resolved it?

**Comment by carher** [comment]
@zeko78 I first tried simply updating to a newer version of AS4-plugin and still running inbound and outbound in the same instance. Took about a day before the problem manifested.

Then we went on to try running on separate instances. We have been doing that for a few hours now, no observed problem yet. But way too soon to conclude.

**Comment by carher** [comment]
@FrodeBjerkholt Note that it is not necessary to have tons of traffic. Some traffic now and then, plus a few breaks, is a better recipe for recreation of the problem, I believe.

**Comment by zeko78** [comment]
@carher Thank you. The separation seems to be working here too, but also just a few hours in..

**Comment by FrodeBjerkholt** [comment]
As @phax commented in issue #69 - This is a bug in WSS4J: https://issues.apache.org/jira/projects/WSS/issues/WSS-660

Until there is a fix in WSS4J, this problem will pervade.

**Comment by gozus19p** [comment]
Hi.
In Intercent-ER PEPPOL access point we set up two different Tomcat instances in this way:

- the first Tomcat instance contains AS2 inbound and AS2+AS4 outbound;
- the second Tomcat instance contains AS4 inbound only.

We added a routing rule on our regional load balancer. This rule basically split AS2 and AS4 traffic between two Tomcat instances making sure that AS2 inbound flow goes into first Tomcat as well as that AS4 flow goes into the second Tomcat instance.

We set up this configuration on the 27th of February. Since that moment we are not experiencing the "No services have been found" issue anymore.
We received around 2.000 documents via AS4 and we never went offline.

If something goes wrong I will post it here, but I do not expect anything extrange. This configuration seems to work fine.

Have a nice day! :)

**Comment by carher** [comment]
I can also report. We have two instances (of Jetty, though), one for inbound (AS2 + AS4) and one for outbound (AS2 + AS4). In this configuration I have not yet seen the original issue I reported, we have been running it like this for 26 days. However, there was a similar issue, but there were simply no logs at all on the inbound instance from that. Currently I'm not able to find the logs from the outbound instance either. We first tried restarting the inbound instance, that did not solve the problem. When we also restarted the outbound instance, communication was okay again. We adjusted log levels after that and hope to know more if it happens again.


Also, I see that issue 72 seems sort of related. Perhaps it will be a good idea to update to version 4.1.6 of the AS4 plugin?
